### PR TITLE
Cyclopedia Character: Send Achievements

### PR DESCRIPTION
**Pull request by omarcopires**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Credits to
@zbizu
@ArturKnopik 
@MillhioreBT
https://github.com/OTAcademy

### Changes Proposed 
image

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Review by EPuncker — DISMISSED** [review]

File: data/scripts/network/cyclopedia_character.lua [hunk 1]
```diff
@@ -24,14 +24,39 @@ local function sendBasicInfo(self, msg)
...
+	local achievementIds = self:getAchievements()
+	msg:addU16(self:getAchievementPoints())
+	msg:addU16(#getSecretAchievements())
+	msg:addU16(#achievementIds)
+
+	for i = 1, #achievementIds do
+		local achievement = achievements[achievementIds[i]]
+		local secret = achievement.secret
+		msg:addU16(achievement.clientId)
+		msg:addU32(os.time()) -- TODO: a new function is needed to add the time the achievement is earned and access it here.
```
Comment (ranisalt) [inline data/scripts/network/cyclopedia_character.lua:37]: What's the value in storage being used for? Maybe we can store the timestamp there.

Comment (EPuncker) [reply data/scripts/network/cyclopedia_character.lua:37]: agreed

Comment (reyaleman) [reply data/scripts/network/cyclopedia_character.lua:37]: it's a boolean value only... timestamp can be stored there

Comment (omarcopires) [reply data/scripts/network/cyclopedia_character.lua:37]: I tried using a new store, but I couldn't import it in the client:

```lua
self:setStorageValue(PlayerStorageKeys.achievementsDateBase + achievement.id, os.time())
```

I really need help with this part, if anyone wants to push it to the bough I'd appreciate it.

Comment (reyaleman) [reply data/scripts/network/cyclopedia_character.lua:37]: You need to replace this line

https://github.com/otland/forgottenserver/blob/becd8551cb1d93e4a9901857196b34285f0baf2d/data/lib/core/achievements.lua#LL704C3-L704C3

with

```lua
self:setStorageValue(PlayerStorageKeys.achievementsDateBase + achievement.id, os.time())
```

Comment (ranisalt) [reply data/scripts/network/cyclopedia_character.lua:37]: > The only issue I'm seeing is that we cannot set a time to achievements that were previously added to a player. 😄

Who said you weren't playing in 01-01-1970 🤷

Comment (omarcopires) [reply data/scripts/network/cyclopedia_character.lua]: @reyaleman i can't test it right now, could you?

File: data/scripts/network/cyclopedia_character.lua [hunk 2]
```diff
@@ -24,14 +24,39 @@ local function sendBasicInfo(self, msg)
...
+	local achievementIds = self:getAchievements()
+	msg:addU16(self:getAchievementPoints())
+	msg:addU16(#getSecretAchievements())
+	msg:addU16(#achievementIds)
+
+	for i = 1, #achievementIds do
+		local achievement = achievements[achievementIds[i]]
+		local secret = achievement.secret
+		msg:addU16(achievement.clientId)
+		msg:addU32(achievement)
```
Comment (reyaleman) [inline data/scripts/network/cyclopedia_character.lua:37]:
```lua
	for _, id in pairs(achievementIds) do
		local achievement = achievements[id]
		local secret = achievement.secret
		msg:addU16(achievement.clientId)
		msg:addU32(self:getStorageValue(PlayerStorageKeys.achievementsBase + id))
```

**Review by EPuncker — APPROVED** [review]

**Review by reyaleman — APPROVED** [review]

**Pull request merged**